src/hotspot/share/cds/aotClassLinker.cpp

```diff
@@ -0,0 +1,292 @@
+#include "memory/resourceArea.hpp"
+#include "oops/constantPool.inline.hpp"
+#include "oops/instanceKlass.hpp"
+#include "oops/klass.inline.hpp"
+#include "runtime/handles.inline.hpp"
+
+AOTClassLinker::ClassesTable* AOTClassLinker::_vm_classes = nullptr;
+AOTClassLinker::ClassesTable* AOTClassLinker::_candidates = nullptr;
+GrowableArrayCHeap<InstanceKlass*, mtClassShared>* AOTClassLinker::_sorted_candidates = nullptr;
+
+#ifdef ASSERT
+bool AOTClassLinker::is_initialized() {
+  assert(CDSConfig::is_dumping_archive(), "AOTClassLinker is for CDS dumping only");
+  return _vm_classes != nullptr;
+}
+#endif
+
+void AOTClassLinker::initialize() {
+  assert(!is_initialized(), "sanity");
+
+  _vm_classes = new (mtClass)ClassesTable();
+  _candidates = new (mtClass)ClassesTable();
+  _sorted_candidates = new GrowableArrayCHeap<InstanceKlass*, mtClassShared>(1000);
+
+  for (auto id : EnumRange<vmClassID>{}) {
+    add_vm_class(vmClasses::klass_at(id));
+  }
+
+  assert(is_initialized(), "sanity");
+
+  AOTConstantPoolResolver::initialize();
+}
+
+void AOTClassLinker::dispose() {
+  assert(is_initialized(), "sanity");
+
+  delete _vm_classes;
+  delete _candidates;
+  delete _sorted_candidates;
+  _vm_classes = nullptr;
+  _candidates = nullptr;
+  _sorted_candidates = nullptr;
+
+  assert(!is_initialized(), "sanity");
+
+  AOTConstantPoolResolver::dispose();
+}
+
+bool AOTClassLinker::is_vm_class(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+  return (_vm_classes->get(ik) != nullptr);
+}
+
+void AOTClassLinker::add_vm_class(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+  bool created;
+  _vm_classes->put_if_absent(ik, &created);
+  if (created) {
+    add_candidate(ik);
+    InstanceKlass* super = ik->java_super();
+    if (super != nullptr) {
+      add_vm_class(super);
+    }
+    Array<InstanceKlass*>* ifs = ik->local_interfaces();
+    for (int i = 0; i < ifs->length(); i++) {
+      add_vm_class(ifs->at(i));
+    }
+  }
+}
+
+bool AOTClassLinker::is_candidate(InstanceKlass* ik) {
+  return (_candidates->get(ik) != nullptr);
+}
+
+void AOTClassLinker::add_candidate(InstanceKlass* ik) {
+  _candidates->put_when_absent(ik, true);
+  _sorted_candidates->append(ik);
+}
+
+bool AOTClassLinker::try_add_candidate(InstanceKlass* ik) {
+  assert(is_initialized(), "sanity");
+
+  if (!CDSConfig::is_dumping_aot_linked_classes() || !SystemDictionaryShared::is_builtin(ik)) {
+    return false;
+  }
+
+  if (is_candidate(ik)) { // already checked.
+    return true;
+  }
+
+  if (ik->is_hidden()) {
+    return false;
+  }
+
+  InstanceKlass* s = ik->java_super();
+  if (s != nullptr && !try_add_candidate(s)) {
+    return false;
+  }
+
+  Array<InstanceKlass*>* interfaces = ik->local_interfaces();
+  int num_interfaces = interfaces->length();
+  for (int index = 0; index < num_interfaces; index++) {
+    InstanceKlass* intf = interfaces->at(index);
+    if (!try_add_candidate(intf)) {
+      return false;
+    }
+  }
```

Comment:
That's possible. Since we go through all the initial set of classes found in `ArchiveBuilder::current()->klasses()`, eventually all classes that are eligible for aot-linking will be added to the candidate list. The reason for the recursion is to

- Filter out classes whose super types are not eligible
- Sort the candidate list so that super types always come before sub types.

Reply:
Doesn't that potentially leave a "super" class that has no subclasses? Or do we not really care?